### ISSUE TITLE
Improve compatibility with tokio

### ISSUE THREAD

**Issue by devrandom** [lightningdevkit/rust-lightning#1367]
### Background

The VLS project is using gRPC to connect to a remote signer.  This uses tokio/tonic.

### Symptom

If the gRPC client is created in the default tokio runtime, a deadlock may result.

### Discussion

The `lightning-net-tokio` calls into `PeerManager` sync code.  This code takes a lock on the `PeerManager.peers` Mutex, which is held across any calls to the signer.

The `ldksample` calls `PeerManager.get_peer_node_ids` in `do_connect_peer`, which also takes a lock on said Mutex.

The above sync code can stall the I/O reactor, resulting in the signer gRPC call never completing and the `peers` Mutex never being released.

A possible solution is to always call `task::block_in_place` or perhaps `Handle::spawn_blocking` when calling any Mutex-using sync code from inside `lightning-net-tokio` and `ldksample`.

**Comment by devrandom** [comment]
On further thought, a more general description of the issue is "holding an `std` `Mutex` across an `await` that performs tokio IO".

I believe that `BitcoindClient` in the LDK sample suffers from this.

The deadlock scenario is:

- thread 1: locks mutex, then starts IO
- thread 2: has the IO reactor, tries to lock mutex

Thread 2 is blocked on the mutex, and therefore the IO reactor is stalled.  Thread 1 is stalled waiting for IO to complete.  No progress can be made - deadlock.

**Comment by TheBlueMatt** [comment]
Right, I believe the issue here is in thread 1 there - we are in a sync context, take a syn mutex, and then move to an async context. In general we should never be going from a sync context to (blocking on) an async context.

**Comment by devrandom** [comment]
Another variation of this is likely the cause of https://github.com/L2-Technology/sensei/pull/73:

- thread 1: takes a lock on sync mutex, then spawns an LDK user task (e.g. persistence) onto thread 2, blocks waiting for task
- thread 2: dispatches an LDK task, which tries to lock same mutex

Thread 1 can't make progress, because the task it is waiting is starved on thread 2.  Thread 2 can't make progress because of the sync lock wait.

My previous conclusion was that it's never safe to spawn user tasks on the runtime LDK uses.  That's why we use the following code to spawn lnrod user tasks in a separate runtime:

https://gitlab.com/lightning-signer/lnrod/-/blob/3e155d2ccecf45212a1bda74e7581a10f0e7ffd5/src/signer/vls.rs#L227-232
